devops: add s390x & ppc64le CI

taronaeo opened the pull request:
closes #13243

Introduce s390x & ppc64le CI using IBM Actions on POWER and Z Runner images.

TODO:
- [x] Fix s390x build warnings (https://github.com/ggml-org/llama.cpp/pull/15928)
- [x] Turn LLAMA_FATAL_WARNINGS back ON
- [x] Fix tokeniser differences between Little-Endian vs. Big-Endian
- [x] (OPTIONAL) Turn on ppc64le CI

Edit: I have to use this PR to test the GitHub Actions as only Llama.cpp is authorised to use the s390x and ppc64le images. Unfortunately I do not have an alternative to develop this elsewhere and submit the PR once I'm ready.

AlekseiNikiforovIBM committed:
Fix unicode flags conversion from and to uint16_t

Bitfields are allocated in different order on s390x

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: move s390x and ppc64le ci build

we have access to ubuntu-24.04-s390x and ppc64le images now

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: disable ppc64le for now since they have compiler errors

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: stop warnings as errors

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: switch to non-macro flag

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: going the llama macro route

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: add big-endian gguf test models

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo commented:
> I think you will need to resolve the warnings and revert before undrafting, the build is done with LLAMA_FATAL_WARNINGS=ON for a reason.

Yep, will take note of that before marking as ready. I wanted to progress past the compile stage to fix the endianness issues with the test models first because that is the problematic part.

Also, the problem about GCC on s390x and ppc64le is that the compiler is a little more strict. Things that would have been okay on x86 or ARM are reported as warnings on s390x and ppc64le and, in this case, marked as fatal.

I suppose we can either have the codeowner work to fix the warning on the affected lines, or (not ideal) we disable warnings as fatal for s390x and ppc64le.

taronaeo committed:
devops: disable ppc64le to test s390x, check test build

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: dup .gguf.inp files for big-endian tests

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: dup .gguf.out files for big-endian too

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

CISC commented:
Looks like there's an endian-issue in `WPM` tokenizer.

taronaeo committed:
devops: add python setup and endian byteswap

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: pooring thing does not have s390x python3

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: add missing rust compiler for s390x

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: try rust actions runner

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
Revert "devops: try rust actions runner"

This reverts commit 3f8db04356033d6c1d7eccc75ca396bc5298250c.

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: try a different path for rust

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: dump home directory and user info

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: install gguf-py only

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: missed relative path

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: remove big-endian files since local swapping is working

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: revert test-tokenizer-0 cmakelists

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

AlekseiNikiforovIBM committed:
Simplify byteswap command

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

AlekseiNikiforovIBM committed:
Add byteswapping and git-lfs for test-tokenizers-ggml-vocabs

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

AlekseiNikiforovIBM committed:
Fix endianness detection in vocab loader

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

AlekseiNikiforovIBM committed:
Disable test-thread-safety on s390x

In this test a model is downloaded,
then immediately loaded to check if more downloads are needed,
and then used for test.

There is no clean way to separate all those steps
 to add byteswapping between them, so just skip this test.

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

AlekseiNikiforovIBM committed:
Fix q8_0 test in test-quantize-fns

vec_signed uses unexpected rounding mode.
Explicitly use different rounding function.

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

CISC commented:
@AlekseiNikiforovIBM Great job on figuring out the endian issues!

It's unfortunate that `unicode_cpt_flags` is implemented using a bitfield struct (it's a little more manageable in C++23, but we're not there yet), it's probably worth reimplementing this into something endian-safe. @ggerganov?

ggerganov commented:
> it's probably worth reimplementing this into something endian-safe. @ggerganov?

Shouldn't be too difficult to do since this struct is never written to disc. So there would be no breaking changes to worry about.

taronaeo committed:
devops: add big-endian stories260K

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: add s390x test-eval-callback

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: fix test does not exist

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: fix model not found llama-eval-callback

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

AlekseiNikiforovIBM committed:
Fix q3_K dot product error in test-quantize-fns on s390x

Array q8bytes had only 4 elements allocated, but 8 elements accessed.
This lead to write out of bounds and later read of overwritten values out of bounds
and incorrect result.

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: re-enable ppc64le for testing

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: activate test-thread-safety for s390x

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: disable ppc64le tests

for some reason it keeps failing test-thread-safety tests and I do not
    have a machine that is able to replicate the tests.

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

taronaeo committed:
devops: LLAMA_FATAL_WARNINGS=ON

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

AlekseiNikiforovIBM committed:
Correct repository URL for s390x for test-thread-safety model

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

CISC commented:
> Any idea why `test-thread-safety` keeps erroring a [segfault](https://github.com/ggml-org/llama.cpp/actions/runs/17975846799/job/51133571282?pr=15925#step:8:26868)? I've tested the same locally on an IBM z17 and there are no errors.

Since there doesn't seem to be any output I'd say it's likely something wrong with the executable, linking issue?

AlekseiNikiforovIBM committed:
Fix fs_get_cache_directory

Ensure it works even if both XDG_CACHE_HOME and HOME are unset.
This might happen in containers.

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

AlekseiNikiforovIBM committed:
Re-enable CI for ppc64le

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

AlekseiNikiforovIBM committed:
Fortify ggml_rope_impl

Only memcpy data from sections argument if it's non-NULL.

Signed-off-by: Aaron Teo <aaron.teo1@ibm.com>

CISC commented on src/unicode.h:
```diff
@@ -34,11 +38,49 @@ struct unicode_cpt_flags {
 
     // decode from uint16
     inline unicode_cpt_flags(const uint16_t flags = 0) {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
```
In `llama-vocab.cpp` these were protected with a `__GNUC__` check, are there any compilers we expect this to fail on?

Also, please add a `TODO` about reimplementing this endian-safe.

AlekseiNikiforovIBM replied:
On the other hand, if __BYTE_ORDER__ and __ORDER_LITTLE_ENDIAN__ are undefined, as with MSVC, it'll hit this case as true and compile little-endian case by default. Additional guard in other code was needed because default case was big-endian. So, additional changes here are not necessary.

CISC commented on tools/mtmd/clip.cpp:
```diff
@@ -3067,7 +3067,7 @@ struct image_manipulation {
 
         float Cc;
-        float C[5];
+        float C[5] = {};
```
I know it's tiny, but please make a separate PR for this. :)

AlekseiNikiforovIBM replied:
Then this PR will depend on that separate PR. As far as I understand, without this change we get warning, which gets treated as an error.

CISC replied:
Yes, so a rebase is required before merge. I just want the visibility.

CISC commented:
> > Any interest in uploading `tinyllamas/stories260K-be.gguf` into the ggml.ai huggingface repository? So that we don't have to point the tests to my personal HuggingFace repository
> 
> Any interest in having the Big-Endian GGUF files located in ggml-org's Hugging Face instead? @ggerganov @slaren

I don't think that would be an issue, the little endian version is already hosted there.

taronaeo commented:
> > > Any interest in uploading `tinyllamas/stories260K-be.gguf` into the ggml.ai huggingface repository? So that we don't have to point the tests to my personal HuggingFace repository
> 
> > 
> 
> > Any interest in having the Big-Endian GGUF files located in ggml-org's Hugging Face instead? @ggerganov @slaren
> 
> 
> 
> I don't think that would be an issue, the little endian version is already hosted there.

Great! Please let me know once the Big-Endian version is up so that we can point the -hf over

ggerganov commented:
@taronaeo It's up: https://huggingface.co/ggml-org/models/tree/main/tinyllamas

AlekseiNikiforovIBM committed:
Add TODO in struct unicode_cpt_flags to reimplement it in endian-independent way

AlekseiNikiforovIBM commented on src/unicode.h:
```diff
@@ -33,12 +37,52 @@ struct unicode_cpt_flags {
     uint16_t is_nfd         : 1;
 
     // decode from uint16
+    // TODO: reimplement this in endian-independent way
```
Sure, I'll update it.

AlekseiNikiforovIBM committed:
Update URL for big-endian model

AlekseiNikiforovIBM commented:
I got rid of commits which added gcc 14 and 15 and then removed them in workflows, merged commits updating big endian detection macros in llama-vocab.cpp and updated rounding builtins, changing parameter from 0 to 4, since other builtins use that mode mostly:

https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/config/s390/vecintrin.h;h=bda8471e5b5e64424bdcdd078776b6d871670129;hb=HEAD#l107

It'd still be great to have `stories15M-be.Q4_0.gguf` model moved into llama.cpp-related location.

AlekseiNikiforovIBM committed:
Update .github/workflows/build.yml

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

AlekseiNikiforovIBM commented:
I've rebased this PR since https://github.com/ggml-org/llama.cpp/pull/16275 was merged, removed miniaudio-related commits since https://github.com/ggml-org/llama.cpp/pull/16212 was merged, and removed commits `devops: add s390x to build-linux-cross` and `Revert "devops: add s390x to build-linux-cross"` since together they change nothing.

AlekseiNikiforovIBM committed:
Update remaining mentions of BE models to ggml-org/models repo

AlekseiNikiforovIBM commented on tests/CMakeLists.txt:
```diff
@@ -185,7 +185,11 @@ llama_build_and_test(test-json-partial.cpp)
+if (NOT ${CMAKE_SYSTEM_PROCESSOR} MATCHES "s390x")
+    llama_build_and_test(test-thread-safety.cpp ARGS -hf ggml-org/models -hff tinyllamas/stories15M-q4_0.gguf -ngl 99 -p "The meaning of life is" -n 128 -c 256 -ub 32 -np 4 -t 2)
+else()
+    llama_build_and_test(test-thread-safety.cpp ARGS -hf taronaeo/tinyllamas-BE -hff stories15M-be.Q4_0.gguf -ngl 99 -p "The meaning of life is" -n 128 -c 256 -ub 32 -np 4 -t 2)
```
Done

CISC commented:
@AlekseiNikiforovIBM Tests passing, wait for @taronaeo to merge or shall I merge when CI is done?

taronaeo commented:
- `CI / ggml-ci-x64-cpu-amx` failing has been the same as other PRs.
- `CI / ggml-ci-arm64-cpu-high-perf-sve`, `CI / ggml-ci-x64-nvidia-vulkan-cm2` and `CI / macOS-latest-cmake-arm64` failing does not seem related to this PR.

`CI / ubuntu-22-cmake-vulkan (pull_request)` This I'm unsure though:
```
	 27 - test-thread-safety (ILLEGAL)                      main
	 29 - test-opt (ILLEGAL)                                main
	 31 - test-backend-ops (ILLEGAL)                        main
	 34 - test-barrier (ILLEGAL)                            main
	 35 - test-quantize-fns (ILLEGAL)                       main
	 36 - test-quantize-perf (ILLEGAL)                      main
	 37 - test-rope (ILLEGAL)                               main
```

CISC commented:
> `CI / ubuntu-22-cmake-vulkan (pull_request)` This I'm unsure though:

It's "OK", it's a corrupt ccache.

taronaeo commented:
> > `CI / ubuntu-22-cmake-vulkan (pull_request)` This I'm unsure though:
> 
> It's "OK", it's a corrupt ccache.

Gotcha. Will proceed to merge then.
